File: swift/Sources/MotionGestureCore/DetectorLifecycle.swift
public enum DetectorLifecycleState: String, Sendable {
  case idle
  case running
  case stopped
}

public enum DetectorOperation: String, Sendable {
  case start
  case consume
  case stop
}

/// Stable invalid-state failure for the detector lifecycle defined by core specification v1.
public struct DetectorInvalidStateError: Error, Equatable, Sendable, CustomStringConvertible {
  public let operation: DetectorOperation
  public let state: DetectorLifecycleState

  public init(operation: DetectorOperation, state: DetectorLifecycleState) {
    self.operation = operation
    self.state = state
  }

  public var code: String { "invalidState" }

  public var description: String {
    "\(code): cannot \(operation.rawValue) while detector is \(state.rawValue)"
  }
}

File: android/motion-gesture-core/src/main/kotlin/io/github/mtatsuto/motiongesture/core/DetectorLifecycle.kt
package io.github.mtatsuto.motiongesture.core

enum class DetectorLifecycleState {
    IDLE,
    RUNNING,
    STOPPED,
}

enum class DetectorOperation {
    START,
    CONSUME,
    STOP,
}

/** Stable invalid-state failure for the detector lifecycle defined by core specification v1. */
class DetectorInvalidStateException(
    val operation: DetectorOperation,
    val state: DetectorLifecycleState,
) : IllegalStateException("invalidState: cannot ${operation.name.lowercase()} while detector is ${state.name.lowercase()}") {
    val code: String = "invalidState"
}
